mergedSchema.js

```diff
@@ -0,0 +1,82 @@
+import { mergeSchemas, introspectSchema, makeRemoteExecutableSchema } from "graphql-tools"
+import { ApolloLink } from "apollo-link"
+import { createHttpLink } from "apollo-link-http"
+import { setContext } from "apollo-link-context"
+import fetch from "node-fetch"
+
+import config from "./config"
+import localSchema from "./schema"
+
+function createConvectionLink() {
+  const httpLink = createHttpLink({
+    fetch,
+    uri: `${process.env.CONVECTION_API_BASE}/graphql`,
+  })
+
+  const middlewareLink = new ApolloLink((operation, forward) => forward(operation))
+
+  const authMiddleware = setContext((_request, context) => {
+    const loaders = context.graphqlContext && context.graphqlContext.res.locals.dataLoaders
+    const tokenLoader = loaders && loaders.convectionTokenLoader
+
+    // If a token loader exists for Convection (i.e. this is an authenticated request), use that token to make
+    // authenticated requests to Convection.
+    if (tokenLoader) {
+      return tokenLoader().then(({ token }) => {
+        return {
+          ...context.headers,
+          headers: {
+            authorization: `Bearer ${token}`,
+          },
+        }
+      })
+    }
+
+    // Otherwise use MP’s XApp token so that on startup MP can fetch Convection’s schema.
+    // TODO: Actually enable once Convection supports this.
+    return {
+      ...context.headers,
+      headers: {
+        // authorization: `XApp ${config.GRAVITY_XAPP_TOKEN}`,
+        authorization: `Bearer ${process.env.CONVECTION_TOKEN}`,
+      },
+    }
+  })
+
+  return middlewareLink.concat(authMiddleware).concat(httpLink)
+}
+
+export default async function mergedSchema() {
+  const convectionLink = createConvectionLink()
+
+  const convectionSchema = await makeRemoteExecutableSchema({
+    schema: await introspectSchema(convectionLink),
+    link: convectionLink,
+  })
+
+  const linkTypeDefs = `
+    extend type Submission {
+      artist: Artist
+    }
+  `
+
+  return mergeSchemas({
```

Comment:
This basically happens once during start up of MP, so one thing to consider is, lets say Convection's GraphLQ schema got updated and there was a new field, we'd ONLY get that update if we restart MP, right?

For now we might be fine with manual restarts of MP, but eventually if this became more problematic (as we add more and more services) it might make sense to use event stream for reloading schema. Each time an app updated it schema we'd refresh this.

Reply:
Correct 👍 And that sounds like a good solution to keep in mind.